Add support for metrics endpoint scrape jobs

fridgepoet opened the pull request:
**What:** This PR adds resources to manage ["metrics endpoint scrape jobs" ](https://grafana.com/docs/grafana-cloud/monitor-infrastructure/integrations/integration-reference/integration-metrics-endpoint/).
Metrics Endpoint enables customers to configure Grafana Cloud managed agents to scrape their publicly addressable metrics endpoint.

**Why:** Setting up these scrape jobs has only been available in the Grafana Cloud UI until now, and this PR provides a solution to manage them via Terraform. ([Design doc](https://docs.google.com/document/d/1YVThjOtBjV0L7IvSlnCjyOVtmCOttLOhCYhiOsDwaps/edit))

**Special notes:** This branch is composed of several separate PRs that have been reviewed individually. It is based on similar work for [AWS scrape jobs](https://github.com/grafana/terraform-provider-grafana/tree/tf-cloud-provider-aws-app). Thank you, Tristan and Pablo.

#### Testing locally

Prerequisite: Follow the [instructions](https://github.com/grafana/terraform-provider-grafana?tab=readme-ov-file#local-development-with-grafana) for local development

1. Create a directory with two files:
* main.tf with the following content
  ```hcl
  provider "grafana" {
    connections_api_url = "<your connections url>"
    connections_api_access_token = "<your access policy token>"
  }
  
  resource "grafana_connections_metrics_endpoint_scrape_job" "test" {
     stack_id = "<your stack id>"
     name = "my_scrape_job"
     enabled = true
     authentication_method = "basic" # or use bearer and authentication_bearer_token
     authentication_basic_username = "<your username>"
     authentication_basic_password = "<your password>"
     url = "<url to scrape>"
     scrape_interval_seconds = 120
   }
  ```
* terraform.tf with the following content
   ```
   terraform {
     required_providers {
       grafana = {
         source = "grafana/grafana"
       }
     }
   }
   ```

2. Run `terraform apply`
3. Navigate to the Metrics Endpoint connection: you should see a successful scrape job created:
![Screenshot 2024-10-29 at 11 01 51](https://github.com/user-attachments/assets/13094e6b-a218-4c57-a19f-1ab208600d21)

fridgepoet committed:
Scaffold metrics endpoint scrape job resource (#1807)

fridgepoet committed:
Add client for Connections API, some wiring up the client to the provider (#1809)

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Merge remote-tracking branch 'origin/main' into HEAD

fridgepoet committed:
Update Connections API client paths (#1827)

fridgepoet committed:
Merge remote-tracking branch 'origin/main' into HEAD

fridgepoet committed:
Fix the expected response from Connections API (#1831)

fridgepoet committed:
Merge remote-tracking branch 'origin/main' into metrics-endpoint

fridgepoet committed:
Implement terraform Resource methods on resourceMetricsEndpointScrapeJob (#1812)

fridgepoet committed:
Merge remote-tracking branch 'origin/main' into metrics-endpoint

fridgepoet committed:
Add metrics endpoint scrape job data source resource (#1817)

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Modify Connections API path for metrics endpoint scrape jobs (#1847)

fridgepoet committed:
Add enabled to metrics endpoint scrape job resource (#1849)

fridgepoet committed:
Merge remote-tracking branch 'origin/main' into metrics-endpoint

matthewnolf commented on examples/resources/grafana_connections_metrics_endpoint_scrape_job/resource.tf:
```diff
@@ -0,0 +1,10 @@
+resource "grafana_connections_metrics_endpoint_scrape_job" "test" {
```
Can we give this a slightly more descriptive name? Like `my-scrape-job`?

matthewnolf commented on examples/resources/grafana_connections_metrics_endpoint_scrape_job/resource.tf:
```diff
@@ -0,0 +1,10 @@
+  authentication_method         = "basic"
+  authentication_basic_username = "my-username"
+  authentication_basic_password = "my-password"
+  url                           = "https://dev.my-metrics-endpoint-url.com:9000/metrics"
```
Can we use a URL grafana owns? Like `grafana.com/metrics`?

matthewnolf commented on internal/common/client.go:
```diff
@@ -14,6 +14,8 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/grafana/terraform-provider-grafana/v3/internal/common/connectionsapi"
```
Is the newline to separate external / internal packages?

fridgepoet replied:
gonna remove the newline

fridgepoet replied:
nevermind I saw your other comment. I don't think there is a rule for the entire repo but I can separate as it was done here. 
I think this particular newline was inadvertent, but I can apply the rule to the imports I'm touching from now on.

matthewnolf commented on internal/common/connectionsapi/client.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+type MetricsEndpointScrapeJob struct {
+	Name                        string `json:"name"`
```
Is `Name` part of the request body? Wondering if we should make it a function argument in the same way we do for `stackID`, e.g.:
`CreateMetricsEndpointScrapeJob(ctx context.Context, stackID string, jobName string, jobData MetricsEndpointScrapeJob)`

fridgepoet replied:
That's right, thanks for pointing that out.

matthewnolf commented on internal/common/connectionsapi/client.go:
```diff
@@ -0,0 +1,146 @@
+	respData := apiResponseWrapper[MetricsEndpointScrapeJob]{}
+	err := c.doAPIRequest(ctx, http.MethodPost, path, &jobData, &respData)
+	if err != nil {
+		return MetricsEndpointScrapeJob{}, fmt.Errorf("failed to create metrics endpoint scrape job: %w", err)
```
With all these error messages, it may be useful to include the job name that failed. 
e.g. `(failed to create metrics endpoint scrape job %s: %w, jobdata.Name, err)`

matthewnolf commented on internal/common/connectionsapi/client.go:
```diff
@@ -0,0 +1,146 @@
+
+var ErrNotFound = fmt.Errorf("not found")
+
+func (c *Client) doAPIRequest(ctx context.Context, method string, path string, body any, responseData any) error {
```
nit: Looks like path never changes, do we need it as a parameter?

matthewnolf replied:
Path does change, so let's leave this.

matthewnolf commented on internal/common/connectionsapi/client.go:
```diff
@@ -0,0 +1,146 @@
+		}
+		reqBodyBytes = bytes.NewReader(bs)
+	}
+	var resp *http.Response
```
nit: Can we declare this where it's used (L128)

matthewnolf commented on internal/common/connectionsapi/client.go:
```diff
@@ -0,0 +1,146 @@
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.authToken))
+	req.Header.Add("Content-Type", "application/json")
```
Does the client already have a user agent header set? If not, can we set one so that it's clear this request is coming from the Terraform provider (and what version)?

fridgepoet replied:
Don't think so but I was able to pass it down from the provider config. Let me know if it looks good to you.

matthewnolf replied:
For the API to attribute the job with terraform, we can use the header we need is `Grafana-Terraform-Provider`

matthewnolf commented on internal/common/connectionsapi/client_test.go:
```diff
@@ -0,0 +1,312 @@
+	"net/http/httptest"
+	"testing"
+
+	"github.com/grafana/terraform-provider-grafana/v3/internal/common/connectionsapi"
```
Can we separate these imports by internal and external packages? Like done in internal/common/client.go

matthewnolf commented on internal/common/connectionsapi/client_test.go:
```diff
@@ -0,0 +1,312 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
```
Can we have a test on the constructor. That it correctly rejects a bad URL

fridgepoet replied:
Yup, added one. Let me know if that's what you meant.

matthewnolf commented on internal/common/connectionsapi/client_test.go:
```diff
@@ -0,0 +1,312 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestClient_sets_auth_token_and_content_type(t *testing.T) {
```
Can we move this into the tests for each operation so that we're testing the package API surface rather than specific elements of functionality

matthewnolf commented on internal/common/connectionsapi/client_test.go:
```diff
@@ -0,0 +1,312 @@
+		}, actualJob)
+	})
+
+	t.Run("returns error when connections API responds 500", func(t *testing.T) {
```
Can we add an error test for when there is an issue connecting with the API? Specifically this error branch https://github.com/grafana/terraform-provider-grafana/pull/1862/files#diff-0fc3e753d1a1d2a0ed9b1b4cd2f6ce527ee0dca118ad383c59c5f6a93bc21242R125

fridgepoet replied:
I am almost able to declare and interface and mock the Do, the only thing is I'm not sure how to set the client Timeout anymore. Let's look at this one together.

matthewnolf commented on internal/resources/connections/data_source_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,110 @@
+		dataTF.Name.ValueString(),
+	)
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to get metrics endpoint scrape job", err.Error())
```
```suggestion
		resp.Diagnostics.AddError("failed to get metrics endpoint scrape job", err.Error())
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+				Required: true,
+			},
+			"authentication_bearer_token": schema.StringAttribute{
+				Description: "Token for authentication bearer.",
```
```suggestion
				Description: "Bearer token used for authentication, use if scrape job is using bearer authentication method",
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+				Optional:    true,
+			},
+			"authentication_basic_username": schema.StringAttribute{
+				Description: "Username for basic authentication.",
```
```suggestion
				Description: "Username for basic authentication, use if scrape job is using basic authentication method",
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+				Optional:    true,
+			},
+			"authentication_basic_password": schema.StringAttribute{
+				Description: "Password for basic authentication.",
```
```suggestion
				Description: "Password for basic authentication, use if scrape job is using basic authentication method",
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+				Optional:    true,
+			},
+			"url": schema.StringAttribute{
+				Description: "The url to scrape metrics; a valid HTTPs URL is required.",
```
```suggestion
				Description: "The url to scrape metrics from; a valid HTTPs URL is required.",
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+	jobClientModel, err := r.client.CreateMetricsEndpointScrapeJob(ctx, dataTF.StackID.ValueString(),
+		convertJobTFModelToClientModel(dataTF))
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to create metrics endpoint scrape job", err.Error())
```
```suggestion
		resp.Diagnostics.AddError("failed to create metrics endpoint scrape job", err.Error())
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+		dataTF.Name.ValueString(),
+	)
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to get metrics endpoint scrape job", err.Error())
```
```suggestion
		resp.Diagnostics.AddError("failed to get metrics endpoint scrape job", err.Error())
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+	jobClientModel, err := r.client.UpdateMetricsEndpointScrapeJob(ctx, dataTF.StackID.ValueString(), dataTF.Name.ValueString(),
+		convertJobTFModelToClientModel(dataTF))
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to update metrics endpoint scrape job", err.Error())
```
```suggestion
		resp.Diagnostics.AddError("failed to update metrics endpoint scrape job", err.Error())
```

matthewnolf commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,257 @@
+		dataTF.Name.ValueString(),
+	)
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to delete metrics endpoint scrape job", err.Error())
```
```suggestion
		resp.Diagnostics.AddError("failed to delete metrics endpoint scrape job", err.Error())
```

matthewnolf commented on internal/resources/examples_test.go:
```diff
@@ -82,6 +82,12 @@ func TestAccExamples(t *testing.T) {
+		{
+			category: "Connections",
+			testCheck: func(t *testing.T, filename string) {
+				t.Skip() // TODO: Make all examples work
```
Can we re-enable this now?

fridgepoet replied:
Let's discuss this one.
Some info:
There is a linter that forces this category to be added, but this test [creates a whole new test](https://github.com/grafana/terraform-provider-grafana/pull/1862/files/22be4c46f85aed7cd6303107de418b701df95683#diff-aff89be1f7ad0485c64ec9853c3fdb51f08cea18faf005a10c6220cd6f976110R121-R131) which cannot pass because we are not depending on a _real_ Connections API/access policy token/real metrics endpoint scrape URL.
So I have matched the t.Skip() of all the tests in other categories in this file. The resource test `TestAcc_MetricsEndpointScrapeJob` still runs elsewhere in the CI.

fridgepoet replied:
Discussed with Matt, we've agreed to t.Skip() as most of the other categories have also done. I've updated the comment to explain why.

matthewnolf commented on pkg/provider/framework_provider.go:
```diff
@@ -40,6 +40,9 @@ type ProviderConfig struct {
 	OncallURL         types.String `tfsdk:"oncall_url"`
 
+	ConnectionsAccessToken types.String `tfsdk:"connections_access_token"`
+	ConnectionsURL         types.String `tfsdk:"connections_url"`
```
Wonder if it's worth calling these `ConnectionsAPI...` instead of just `Connections` which may be slightly vague?

matthewnolf commented on pkg/provider/framework_provider.go:
```diff
@@ -191,6 +196,16 @@ func (p *frameworkProvider) Schema(_ context.Context, _ provider.SchemaRequest,
+			},
+			"connections_url": schema.StringAttribute{
+				Optional:            true,
+				MarkdownDescription: "A Grafana Connections API backend address. May alternatively be set via the `GRAFANA_CONNECTIONS_URL` environment variable.",
```
```suggestion
				MarkdownDescription: "A Grafana Connections API address. May alternatively be set via the `GRAFANA_CONNECTIONS_URL` environment variable.",
```

matthewnolf commented on pkg/provider/legacy_provider.go:
```diff
@@ -136,6 +136,18 @@ func Provider(version string) *schema.Provider {
+			"connections_url": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "A Grafana Connections API backend address. May alternatively be set via the `GRAFANA_CONNECTIONS_URL` environment variable.",
```
```suggestion
				Description:  "A Grafana Connections API address. May alternatively be set via the `GRAFANA_CONNECTIONS_URL` environment variable.",
```

matthewnolf commented on internal/resources/connections/metrics_endpoint_scrape_job_test.go:
```diff
@@ -0,0 +1,175 @@
+	})
+}
+
+func Test_httpsURLValidator(t *testing.T) {
```
Can we put this in a resource test file? Seeing as it's involved with the create resource flow

matthewnolf reviewed (commented):
Great work!

cristiangreco commented on internal/resources/examples_test.go:
```diff
@@ -82,6 +82,12 @@ func TestAccExamples(t *testing.T) {
+		{
+			category: "Connections",
+			testCheck: func(t *testing.T, filename string) {
+				t.Skip() // TODO: Make all examples work
```
Should this be updated now?

fridgepoet replied:
Matt asked the same question, the short answer is No and every other category is also skipped. https://github.com/grafana/terraform-provider-grafana/pull/1862#discussion_r1820519059

fridgepoet committed:
Update internal/resources/connections/data_source_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update pkg/provider/legacy_provider.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Changes from review

fridgepoet committed:
Test headers for each client method

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Update internal/resources/connections/resource_metrics_endpoint_scrape_job.go

Co-authored-by: MattNolf <86960799+matthewnolf@users.noreply.github.com>

fridgepoet committed:
Change naming of Connections to Connections API

fridgepoet committed:
Move httpValidator and tests to resources files

fridgepoet committed:
Generate docs

fridgepoet committed:
Metrics endpoint scrape jobs: add docs (#1861)

fridgepoet committed:
Add tests after review

fridgepoet committed:
Merge remote-tracking branch 'origin/metrics-endpoint' into metrics-endpoint

fridgepoet committed:
Rename Connections to Connections API

fridgepoet committed:
Update examples test skip comment

tristanburgess commented on docs/index.md:
```diff
@@ -224,6 +226,57 @@ resource "grafana_oncall_escalation" "example_notify_step" {
+
+Also, by default the Access Policies UI will not show those scopes, to find name you need to use the `Add Scope` textbox, as shown in the following image:
+
+<img src="https://grafana.com/media/docs/grafana-cloud/aws/cloud-provider-terraform-access-policy-creation.png" width="700"/>
```
Should this use a separate screenshot instead of reusing the cloud provider one? Just thinking that the name of the access token is called `cloud-provider-terraform`. We may not want to encourage people to use the same access token for both, in case our needs ever diverge permissions-wise.

matthewnolf replied:
Yeah makes sense, this came up on the PR for it https://github.com/grafana/terraform-provider-grafana/pull/1861#discussion_r1819165332

Do you have guidance on where to host?

tristanburgess replied:
I'm not sure where @thepalbi found `https://grafana.com/media/docs/grafana-cloud/aws/` and how that got setup. If there isn't a similar folder for Connections, maybe good to create one. Otherwise I've seen searching in the Grafana org that plenty of screenshots exist just in `https://grafana.com/media/docs/grafana-cloud/` https://github.com/search?q=org%3Agrafana+%22media%2Fdocs%2Fgrafana-cloud%2F%22&type=code

matthewnolf replied:
Updated in 791c7be017d73b48b4ef3663b3a87c4baaf9ff90

tristanburgess commented on docs/index.md:
````diff
@@ -224,6 +226,57 @@ resource "grafana_oncall_escalation" "example_notify_step" {
+}
+```
+
 ## Authentication
````
Can we add `connections_api_access_token` to this section, as was done for https://github.com/grafana/terraform-provider-grafana/blob/6e384ec04b41e7dd77b8aaa5978125721fd3b763/docs/index.md?plain=1#L380

matthewnolf replied:
Added in 88c37be89d0be8b564bf3000b54b11582bb5672f

tristanburgess commented on internal/resources/connections/resource_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,210 @@
+				Computed:    true,
+				Default:     booldefault.StaticBool(true),
+			},
+			"authentication_method": schema.StringAttribute{
```
is this field necessary? I think you could do a resource-level validation to ensure that either both `authentication_bearer_token` and `authentication_basic_username` are set, or `authentication_basic_password` is set, but not both kinds. See https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators@v0.14.0/resourcevalidator. These would help to catch any misconfigurations at plan time

matthewnolf replied:
Thanks for the suggestion, I've added in resource validation here aa0e52d12e5265f74c7d8e1c9a5d30afc3b47e86 so that we can catch these misconfigurations at plan time.

Re is the field necessary - the API does in deed expect it. If we remove it from the schema, we'd have to add some more logic in the client to re-populate it. Ultimately this is moving us away from our goal of having the terraform implementation a thin layer on top of our API spec without any bespoke logic.
I think in the future, we can look at removing this field from the job API entirely, and then subsequently from here - but think the validation resolves many of the possible paper cuts here for now. 
Let me know if it makes sense to you.

tristanburgess replied:
I'll ultimately leave the decision up to you in terms of your goals etc. My suggestion came from perceiving some overlap in the fields that a user must configure to use the provider. Removing overlap on the Terraform side would make user configuration simpler at the cost of, as you say, complicating the provider implementation vs your goal of implementing a thin layer over the API. It sounds like the validation suggestion still benefits, which is great, and in the future, as you say, you may just remove the field from the API altogether, which sounds like a good plan.

matthewnolf replied:
Will resolve this and leave for a followup action as we can take this out without it being a breaking change.

tristanburgess commented on internal/resources/connections/data_source_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,111 @@
+	)
+}
+
+func (r *datasourceMetricsEndpointScrapeJob) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
```
similar comment as earlier for the datasource and resource: most of the pointer receivers can become value receivers. The only exception is this `Configure` function which does modify the datasource/resource's client

matthewnolf replied:
We've followed the convention to not mix pointer and value receivers for the type (`datasourceMetricsEndpointScrapeJob`). Because `Configure` needs to be a receiver, we've kept them all consistent.

tristanburgess replied:
oh sorry, I deleted a different comment I left on this PR due to your reasoning, and forgot I left this one on here. Thanks for pointing this out

matthewnolf commented on pkg/provider/configure_clients.go:
```diff
@@ -172,6 +179,20 @@ func createOnCallClient(providerConfig ProviderConfig) (*onCallAPI.Client, error
 	return onCallAPI.New(providerConfig.OncallURL.ValueString(), providerConfig.OncallAccessToken.ValueString())
 }
 
+func createConnectionsClient(client *common.Client, providerConfig ProviderConfig) error {
```
We should add support for custom headers to be injected, such as`Grafana-Terraform-Provider`. As an example: https://github.com/grafana/terraform-provider-grafana/pull/1828/files

matthewnolf replied:
Added in 7ff3e9c0e6bab52aa90dd2b22bbbf9f13900696e

matthewnolf committed:
Add auth docs section for connections_api_access_token (#1871)

This change adds a small section to the main docs 'Authentication' subheading, describing the connections_api_access_token and how to provision it.

matthewnolf commented on internal/resources/connections/data_source_metrics_endpoint_scrape_job.go:
```diff
@@ -0,0 +1,111 @@
+				Description: "Method to pass authentication credentials: basic or bearer.",
+				Computed:    true,
+			},
+			"authentication_bearer_token": schema.StringAttribute{
```
We're not reading these are we? Does it make sense for them to be included in the schema?

matthewnolf committed:
Add ConfigValidators for bearer and basic auth combinations (#1872)

The Metrics Endpoint Job type accepts either basic or bearer auth types.
This change adds validation so that either basic or bearer auth is set on the resource, but not both.
It also adds a check that if basic username is provided, then basic password must also be set.

matthewnolf committed:
Inject provider headers into connections client (#1875)

This adds defaultHeaders to the connections client, which are set for each http request.

These headers are set by the provider, during instantiation of the connections client, and enable propagation of common client headers

matthewnolf committed:
Update docs screenshot URL (#1876)

This change updates the URL to point to a specific screenshot for configuring a connections access policy.

matthewnolf committed:
Handle non 200 error codes from connections API specifically (#1878)

* Handle 401 error codes from connections API

* Remove error body from message due to verbosity

matthewnolf committed:
Adds a codeowner for connections resources (#1879)

matthewnolf committed:
Merge branch 'main' into metrics-endpoint

tristanburgess reviewed (approved):
looks good to me! 🚀 :shipit: 🚀

cristiangreco commented on templates/index.md.tmpl:
```diff
@@ -32,6 +32,57 @@ The changelog for this provider can be found here: <https://github.com/grafana/t
+[These docs](https://grafana.com/docs/grafana-cloud/account-management/authentication-and-permissions/access-policies/authorize-services/#create-an-access-policy-for-a-stack) will guide you on how to create
+an access policy. The required permissions, or scopes, are `integration-management:read`, `integration-management:write` and `stacks:read`.
+
+Also, by default the Access Policies UI will not show those scopes, to find name you need to use the `Add Scope` textbox, as shown in the following image:
```
"to find name" reads a bit odd to me? (not a native speaker)

matthewnolf replied:
Updated grammar

matthewnolf committed:
Update grammer

matthewnolf committed:
Merge branch 'main' into metrics-endpoint
